### PR TITLE
policy: Extract common utilities into `kubert`

### PR DESCRIPTION
**Pull request by olix0r**
[kubert] is a library that extracts a bunch of our common
controller utilities into a reusable crate, including:

* Kubernetes client initialization with CLI-specified context (for local
  development)
* Logging initialization, with plain or JSON formatting
* An HTTPS server that reloads its certificates on each connection (for
  admission controllers)
* Signal handling and graceful shutdown
* A simple admin server with `/live` and `/ready` endpoints

This change removes this logic from the `policy-controller` in favor
of the reusable utilities.

[kubert]: https://github.com/olix0r/kubert

### CODE REVIEW

**Comment by olix0r** [comment]
@hawkw https://github.com/linkerd/linkerd2/pull/7922/files#diff-57fc7cddad7686e3e98e2af8114c4939140164540818b8d3a4c389199f8f8160R257-R259 -- cli flag changes

**Comment by hawkw** [comment]
> @hawkw https://github.com/linkerd/linkerd2/pull/7922/files#diff-57fc7cddad7686e3e98e2af8114c4939140164540818b8d3a4c389199f8f8160R257-R259 -- cli flag changes

oh, got it!

File: policy-controller/src/main.rs [hunk 1]
```diff
@@ -117,15 +113,32 @@ async fn main() -> Result<()> {
...
+            .spawn(admission::Service { client }, drain_rx.clone())
+            .await?;
+        info!(addr = %srv.local_addr(), "Admission controller server listening");
     }
 
+    // Spawn a task that emits a log message when shutdown starts.
+    tokio::spawn(async move {
+        let release = drain_rx.signaled().await;
+        info!("Shutdown signaled");
+        drop(release);
```
[the remark this answers is not in the document]
Comment (hawkw) [reply policy-controller/src/main.rs:131]: i believe this is just to make that more obvious to the reader.

[the remark this answers is not in the document]
Comment (olix0r) [reply policy-controller/src/main.rs:131]: Correct. When we explicitly use `drop`, it's generally to signal that the drop does something semantically important. In this case, the drop releases the shutdown handle (in part -- all other drain handles need to be released, too).

File: policy-controller/src/admission.rs [hunk 2]
```diff
@@ -1,24 +1,41 @@
...
 pub struct Service {
     pub client: kube::Client,
 }
 
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("failed to read request body: {0}")]
+    Request(
+        #[from]
+        #[source]
```
Comment (hawkw) [inline policy-controller/src/admission.rs:21]: nit: i don't believe we need both of these; per the `thiserror` README, 
>   The `#[from]` attribute always implies that the same field is `#[source]`, so
  you don't ever need to specify both attributes.
```suggestion
        #[from]
```

File: policy-controller/src/admission.rs [hunk 3]
```diff
@@ -1,24 +1,41 @@
...
+        hyper::Error,
+    ),
+
+    #[error("failed to encode json response: {0}")]
+    Json(
+        #[from]
+        #[source]
```
Comment (hawkw) [inline policy-controller/src/admission.rs:28]:
```suggestion
        #[from]
```

File: policy-controller/src/main.rs [hunk 4]
```diff
@@ -110,22 +106,41 @@ async fn main() -> Result<()> {
...
-        let (listen_addr, serve) = bind_admission_controller(bind_addr, client).await?;
-        tokio::spawn(serve.instrument(info_span!("admission")));
-        info!(addr = %listen_addr, "Admission controller server listening");
+    if admission_controller_disabled {
+        tracing::info!("Admission controller disabled");
+    } else {
+        let srv = server
+            .spawn(admission::Service { client }, drain_rx.clone())
+            .await?;
+        info!(addr = %srv.local_addr(), "Admission controller server listening");
```
Comment (hawkw) [inline policy-controller/src/main.rs:122]: what does the future returned by `kubert::server::ServerArgs::spawn` actually represent? does that mean the server is running, or the server task has _completed_? this was not super obvious from the docs --- I'm guessing it means the server is running, but this code looked a little sketchy the first time i saw it, and i wanted to make sure this wasn't accidentally awaiting the server task _ending_

**Review by hawkw — APPROVED** [review]
lgtm overall!

File: policy-controller/src/main.rs (hunk 4 above)
Comment (olix0r) [reply policy-controller/src/main.rs:122]: yeah the binding happens async and can fail before the server task is actually spawned, I'll document spawn.

**Pull request closed without merging**